torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -0,0 +1,638 @@
+        bucket_cap_kb (int):
+            the size of the buffer used to batch the small parameter tensors, in kilobytes (default 128k).
+    """
+
+    def __init__(
+        self,
+        params: _params_t,
+        optim: Type[Optimizer],
+        group: Optional[Any] = None,
+        bucket_cap_kb: int = 2 ** 17,
+        **default: Any,
+    ):
+        # Hold all the model params in the root .param_groups
+        self.in_super_constructor = True
+        super().__init__(params, default)
+        self.in_super_constructor = False
+
+        # Partition information. lazy evaluation, computed if requested
+        self._per_device_params: OrderedDict[
+            torch.device, List[List[Parameter]]
+        ] = OrderedDict()  # device, rank, params
+        self._param_rank: Dict[torch.Tensor, int] = {}
+        self._partition_parameters: List[List[Dict]] = []
+
+        # Build the wrapped optimizer, responsible for a shard of the params
+        self.group = group if group is not None else dist.group.WORLD
+        self.world_size = dist.get_world_size(self.group)
+        self.rank = dist.get_rank(self.group)
+        self.global_rank = _get_global_rank(self.group, self.rank)
+
+        self.optim = optim(self.partition_parameters()[self.rank], **default)
+
+        # - Sync local and global param_groups keys
+        for global_group, local_group in zip(self.param_groups, self.optim.param_groups):
+            for k, v in local_group.items():
+                if k != "params":
+                    global_group[k] = v
+
+        #  Optional consolidated optimizer state
+        self._all_states: List[Dict[str, Any]] = []
+
+        # Current default device is set by the parameters allocated to this rank
+        self._device = self.partition_parameters()[self.rank][0]["params"][0].device
+        self._buckets: Dict[torch.device, List[Bucket]] = {}
+        for device, per_device in self.per_device_params.items():
+            # Allocate one buffer per rank and per device to group the small parameters
+            self._buckets[device] = [
+                Bucket(buffer=torch.zeros(broadcast_buffer_size, dtype=per_device[0][0].dtype, device=device))
+                for _ in range(len(per_device))
+            ]
+
+    def clip_grad_norm(self, max_norm: Union[float, int], norm_type: Union[float, int] = 2.0) -> torch.Tensor:
+        """
+        Clip all gradients at this point in time. The norm is computed over all gradients together, as if they were
+        concatenated into a single vector. Gradients are modified in-place.
+        Arguments:
+            max_norm (float or int): max norm of the gradients
+            norm_type (float or int): type of the used p-norm. Can be ``'inf'`` for infinity norm.
+        Returns:
+            Total norm of the parameters (viewed as a single vector).
+        .. note: This is analogous to `torch.nn.utils.clip_grad_norm_` but handles the partitioning and multiple
```

Comment:
I think it's something to discuss really, with a larger audience, because if gradient sharding is not to be supported then this is not very impactful and breaks compatibility with the upper layers (fairscale, etc..). 

Happy to relocate the method somewhere else, but the logic goes:
- zero redundancy optimizer makes sure that one rank works on a unique shard of the model
- this in turn means that there's no point in sending all the grads to all the ranks, sure you can do that but probably that this will not be done (at least it makes sense not to do that, and empirically many people are interested in not doing that, ie: get free RAM)
- in turns it means that *if* you specialized your grad communication pattern, the torch clip util does not work (it will be subtly wrong, which is even worse). 
- hence this util, which is being used for real 

It could be placed anywhere, it just needs to know the partition, same as the optimized DDP. 

I think that if this PR breaks all paths towards a proper Zero then it's not that useful to be honest, something like this will need to be added again, better wait and do it properly ?

Cc @mrshenli @msbaines for diverging opinions